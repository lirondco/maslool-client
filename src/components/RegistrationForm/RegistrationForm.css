@media screen and (min-width: 500px) {
form {
    width: 80% !important;
    margin: 0 auto !important;
  }
  
  label,
  input {
    display: inline-block !important;
  }
  
  label {
    width: 30% !important;
    text-align: right !important;
  }
  
  label+input {
    width: 30% !important;
    margin: 1% 30% 0 4% !important;
  }
    
  input+input {
    float: right !important;
  }

  button {
    display: inline-block !important;
    margin: 1% 40% 1% 20% !important;
  }

  form a, form .error {
      margin: 1% 40% 0 20% !important;
  }

  form a {
      text-decoration: underline !important;
      color: blue;
  }

  form a:hover {
      color: purple;
  }
  
  form a:active {
      color: purple;
  }
}

@media screen and (max-width: 499px) {

form {
    display: flex;
    flex-direction: column;
    margin: 3%;
}

form .main_reg_form {
    display: flex;
    flex-direction: column;
    margin: 1%;
}

form input, form label {
    margin: 1%;
}

form button {
    margin: 2%;
}

form a, form .error {
    margin: 2%;
}

form a {
    text-decoration: underline;
    color: blue;
}

form a:hover {
    color: purple;
}

form a:active {
    color: purple;
}
}