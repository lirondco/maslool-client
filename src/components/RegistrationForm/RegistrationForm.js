import React, { Component } from "react";
import { Link } from "react-router-dom";
import { Input, Required, Label, Button } from "../Utils/Utils";
import AuthApiService from "../../services/auth-api-service";
import "./RegistrationForm.css";
import UserContext from "../../contexts/UserContext";
import Loading from "../Loading/Loading";

export default class RegistrationForm extends Component {
  static defaultProps = {
    onRegistrationSuccess: () => {},
  };
  state = {
    error: null,
    isLoading: false,
    password: null,
    verify: null,
    isMatch: false,
  };

  static contextType = UserContext;

  firstInput = React.createRef();

  handlePasswordChange = (ev) => {
    this.setState({ password: ev.target.value });
    if (this.state.verify !== ev.target.value) {
      this.setState({ isMatch: false, error: "Password must match" });
    } else {
      this.setState({ error: null, isMatch: true });
    }
  };

  handlePasswordCheck = (ev) => {
    this.setState({ verify: ev.target.value });
    if (ev.target.value !== this.state.password) {
      this.setState({ isMatch: false, error: "Password must match" });
    } else {
      this.setState({ error: null, isMatch: true });
    }
  };

  handleSubmit = (ev) => {
    ev.preventDefault();
    this.setState({ error: "Processing your request..." });
    const { username, password, email } = ev.target;
    AuthApiService.postUser({
      username: username.value,
      password: password.value,
      email: email.value,
    })
      .then((user) => {
        this.setState({ error: null, isLoading: true, password: null });
        AuthApiService.postLogin({
          username: username.value,
          password: password.value,
        }).then((res) => {
          username.value = "";
          password.value = "";
          email.value = "";
          this.context.processLogin(res.authToken);
          this.props.onRegistrationSuccess();
          this.setState({ isLoading: false });
        });
      })
      .catch((res) => {
        this.setState({ error: res.error });
      });
  };

  componentDidMount() {
    this.firstInput.current.focus();
  }

  componentWillUnmount() {
    this.setState({ password: null });
  }

  render() {
    const { error } = this.state;
    if (this.state.isLoading === true) {
      return <Loading />;
    }
    return (
      <>
        <h2>Register</h2>
        <hr />
        <form onSubmit={this.handleSubmit}>
          <div className="main_reg_form">
            <Label htmlFor="registration-username-input">
              Username: <Required />
            </Label>
            <Input
              ref={this.firstInput}
              id="registration-username-input"
              name="username"
              required
            />
            <Label htmlFor="registration-password-input">
              Password: <Required />
            </Label>
            <Input
              onChange={this.handlePasswordChange}
              id="registration-password-input"
              name="password"
              type="password"
              required
            />
            <Label htmlFor="registration-verify-password-input">
              Repeat Password: <Required />
            </Label>
            <Input
              onChange={this.handlePasswordCheck}
              type="password"
              id="registration-verify-password-input"
              name="verify-password"
              required
            />
            <Label htmlFor="registration-email-input">
              Email Address: <Required />
            </Label>
            <Input id="registration-email-input" name="email" required />
          </div>
          <div role="alert">{error && <p className="error">{error}</p>}</div>
          <footer>
            <Button disabled={this.state.isMatch === false} type="submit">
              Register
            </Button>{" "}
            <br />
            <Link to="/login">Already have an account?</Link>
          </footer>
        </form>
      </>
    );
  }
}
