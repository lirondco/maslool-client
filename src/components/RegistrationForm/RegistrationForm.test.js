import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import RegistrationForm from "./RegistrationForm";
import { withContext } from "shallow-with-context";

describe(`RegistrationForm component`, () => {

  const context = {
    processLogin: () => {}
  };
  const ComponentWithContext = withContext(RegistrationForm, context);

  it("renders the form", () => {
    const wrapper = shallow(<ComponentWithContext />, { context } );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("form submits when submitted", () => {
    const wrapper = shallow(<ComponentWithContext />, { context });
    const mockedEvent = {
      preventDefault: () => {},
      target: { username: { value: "Test" }, password: { value: "test" }, email: { value: 'test@ema.il' } },
    };
    wrapper.find("form").at(0).simulate("submit", mockedEvent);
  });
});
