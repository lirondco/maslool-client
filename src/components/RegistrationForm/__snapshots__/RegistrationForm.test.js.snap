// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegistrationForm component renders the form 1`] = `
<Fragment>
  <h2>
    Register
  </h2>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <div
      className="main_reg_form"
    >
      <Label
        htmlFor="registration-username-input"
      >
        Username: 
        <Required />
      </Label>
      <ForwardRef
        id="registration-username-input"
        name="username"
        required={true}
      />
      <Label
        htmlFor="registration-password-input"
      >
        Password: 
        <Required />
      </Label>
      <ForwardRef
        id="registration-password-input"
        name="password"
        onChange={[Function]}
        required={true}
        type="password"
      />
      <Label
        htmlFor="registration-verify-password-input"
      >
        Repeat Password: 
        <Required />
      </Label>
      <ForwardRef
        id="registration-verify-password-input"
        name="verify-password"
        onChange={[Function]}
        required={true}
        type="password"
      />
      <Label
        htmlFor="registration-email-input"
      >
        Email Address: 
        <Required />
      </Label>
      <ForwardRef
        id="registration-email-input"
        name="email"
        required={true}
      />
    </div>
    <div
      role="alert"
    />
    <footer>
      <ForwardRef
        disabled={true}
        type="submit"
      >
        Register
      </ForwardRef>
       
      <br />
      <Link
        to="/login"
      >
        Already have an account?
      </Link>
    </footer>
  </form>
</Fragment>
`;
